api/envoy/service/ext_proc/v3/external_processor.proto

```diff
@@ -167,30 +167,31 @@ message ProcessingRequest {
 // [#next-free-field: 11]
 message ProcessingResponse {
   // The response type that is sent by the server.
+  // If the server does not set any of the response type, the data plane will interpret
+  // it as the server has no mutation for the request message.
   oneof response {
-    option (validate.required) = true;
 
-    // The server must send back this message in response to a message with the
+    // The server might send back this message in response to a message with the
```

Comment:
Another high-level question:
Was there any internal validation by Envoy that the response has one of these fields populated? What did it do if no field was set? What did it do if more than one field was set?
If the proposed change creates a breaking change, would it be possible to support both the "old" (must) and the "new" (might) use-cases and add some flag that the ext-proc will send to the Envoy and tell it how to ingest the response?

Reply:
As this is oneof field, so more than one field set is not possible.

If no field is set, there is a check in Envoy which claim the response is unhandled response: https://github.com/envoyproxy/envoy/blob/c352e1c0937dd5594a5d0689641a6fe26e0bf465/source/extensions/filters/http/ext_proc/ext_proc.cc#L1815, and send a local reply for the main-stream request if fail-close is configured: https://github.com/envoyproxy/envoy/blob/c352e1c0937dd5594a5d0689641a6fe26e0bf465/source/extensions/filters/http/ext_proc/ext_proc.cc#L1838.

This API change is indeed a breaking change for customers which expect Envoy to send a local reply for empty response from ext_proc servers.

Reply:
The idea of this change is to make Envoy more tolerant to some "careless" ext_proc servers who does not follow the old API. Supporting both "old" and "new" behavior might be challenging.